File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.kt
/*
괄호 변환 / 코틀린

(,)갯수 같 -> 균현잡힌 괄호 문자열
(,) 짝이 맞 -> 올바른 괄호 문자열
*/
class Solution {
    fun makeString(s : String) : String {
        if(s.isEmpty()) return ""

        var cnt : Int = 0
        var left : Int = 0
        var u : String = ""
        var v : String = ""
        var flag : Boolean = true

        for(i in s){
            if(i == '(') left++ else left --
            cnt++
            if(left == 0){
                u = s.substring(0,cnt)
                v = s.substring(cnt, s.length)
                break
            }
        }

        cnt = 0
        for(i in u){
            if(i == '(') cnt++ else cnt--
            if(cnt < 0) flag = false
        }
        if(flag){
            return u + makeString(v)
        }else{
            var first = "(" +makeString(v) + ")"
            var second = ""
            u.substring(1, u.length - 1).forEach {
                if(it == '(') second = "$second)"
                else second = "$second("
            }
            return first + second
        }
    }

    fun solution(p: String): String {
        var answer = ""
        answer = makeString(p)
        return answer
    }
}

File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.cpp
/*
괄호 변환

(,)의 갯수가 같다면 -> 균현잡힌 문자열
짝이 모두 맞다면 -> 올바른 괄호 문자ㅕㅇㄹ

u,v분리
*/
#include <string>
#include <vector>

using namespace std;

bool isRightString(string s){
    int tmp = 0;
    for(char c : s) {
        if(c == '(')
            tmp++;
        if(c == ')')
            tmp --;
        if(tmp < 0)
            return false;
    }
    return true;
}

string solution(string p) {
    string answer = "";
    if(p == "")
        return "";
    
    string u, v;
    
    int count = 0;
    for(int i = 0 ; i < p.length(); i++){
        if(p[i] == '(')
            count++;
        if(p[i] == ')')
            count--;
        
        if(count == 0){
            u = p.substr(0,i + 1);
            v = p.substr(i + 1);
            break;
        }
    }
    
    if(isRightString(u)){
        answer = u + solution(v);
    }else {
        string tmp = "";
        for(int i = 1; i <= u.length() - 2; i++){
            if(u[i] == '('){
                tmp += ")";
            }else {
                tmp += "(";
            }
        }
        
        answer = '(' + solution(v) + ')' + tmp;
    }
    
    
    return answer;
}
